iceoryx_posh/include/iceoryx_posh/internal/popo/base_client.hpp

```diff
@@ -0,0 +1,148 @@
+    ///
+    bool hasResponses() const noexcept;
+
+    ///
+    /// @brief Check if response has been missed since the last call of this method.
+    /// @return True if response has been missed.
+    /// @details Response may be missed due to overflowing receive queue.
+    ///
+    bool hasMissedResponses() noexcept;
+
+    /// @brief Releases any unread queued response.
+    void releaseQueuedResponses() noexcept;
+
+    friend class NotificationAttorney;
+
+  protected:
+    using SelfType = BaseClient<Port>;
+    using PortType = Port;
+
+    BaseClient(const capro::ServiceDescription& service, const ClientOptions& clientOptions) noexcept;
+
+    /// @brief Only usable by the WaitSet/Listener, not for public use. Invalidates the internal triggerHandle.
+    /// @param[in] uniqueTriggerId the id of the corresponding trigger
+    void invalidateTrigger(const uint64_t uniqueTriggerId) noexcept;
+
+    /// @brief Only usable by the WaitSet/Listener, not for public use. Attaches the triggerHandle to the internal
+    /// trigger.
+    /// @param[in] triggerHandle rvalue reference to the triggerHandle. This class takes the ownership of that handle.
+    /// @param[in] clientState the state which should be attached
+    void enableState(iox::popo::TriggerHandle&& triggerHandle, const ClientState clientState) noexcept;
+
+    /// @brief Only usable by the WaitSet/Listener, not for public use. Returns method pointer to the event
+    /// corresponding hasTriggered method callback
+    /// @param[in] clientState the state to which the hasTriggeredCallback is required
+    WaitSetIsConditionSatisfiedCallback
+    getCallbackForIsStateConditionSatisfied(const ClientState clientState) const noexcept;
+
+    /// @brief Only usable by the WaitSet/Listener, not for public use. Resets the internal triggerHandle
+    /// @param[in] clientState the state which should be detached
+    void disableState(const ClientState clientState) noexcept;
+
+    /// @brief Only usable by the WaitSet/Listener, not for public use. Attaches the triggerHandle to the internal
+    /// trigger.
+    /// @param[in] triggerHandle rvalue reference to the triggerHandle. This class takes the ownership of that handle.
+    /// @param[in] clientEvent the event which should be attached
+    void enableEvent(iox::popo::TriggerHandle&& triggerHandle, const ClientEvent clientEvent) noexcept;
+
+    /// @brief Only usable by the WaitSet/Listener, not for public use. Resets the internal triggerHandle
+    /// @param[in] clientEvent the event which should be detached
+    void disableEvent(const ClientEvent clientEvent) noexcept;
+
+    ///
+    /// @brief const accessor of the underlying port
+    ///
+    const Port& port() const noexcept;
+
+    ///
+    /// @brief accessor of the underlying port
+    ///
+    Port& port() noexcept;
+
+  protected:
+    Port m_port{nullptr};    // NOLINT(cppcoreguidelines-non-private-member-variables-in-classes)
```

Comment:
Let's assume `Port=ClientPortUser` , like the default template value for that argument.

`ClientPortUser` has a constructor which takes a `MemberType_t = ClientPortData` mutable reference - how can you initialize this with a `nullptr`? Why does this compile? I am confused and maybe I am a little bit lost in the valley of templates.

Reply:
You are correct, this will never be a `nullptr` but I just initialized it in case someone adds a new ctor and forgets to set `m_port`

Reply:
@elfenpiff oh, now I understand what you mean ... it works because it is initialized in the cTor therefore the initialization with the `nullptr` never happens. I remove the initialization since it is not required and adding a new cTor will require to also initialize this member. I copied it from the `Publisher`/`Subscriber` which allow a initialization with a nullptr for testing but for `Client`/`Server` I want to use a different approach which does not require to have an invalid port.